contracts/cw20-merkle-airdrop/src/contract.rs

```diff
@@ -0,0 +1,508 @@
+    }
+
+    // check merkle root length
+    let mut root_buf: [u8; 32] = [0; 32];
+    hex::decode_to_slice(merkle_root.to_string(), &mut root_buf)?;
+
+    let latest_stage: u8 = STAGE.load(deps.storage)?;
+    let stage = latest_stage + 1;
+
+    MERKLE_ROOT.save(deps.storage, U8Key::from(stage), &merkle_root)?;
+    STAGE.save(deps.storage, &stage)?;
+
+    Ok(Response {
+        messages: vec![],
+        attributes: vec![
+            attr("action", "register_merkle_root"),
+            attr("stage", stage),
+            attr("merkle_root", merkle_root),
+        ],
+        data: None,
+        events: vec![],
+    })
+}
+
+pub fn execute_claim(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    stage: u8,
+    amount: Uint128,
+    proof: Vec<String>,
+) -> Result<Response, ContractError> {
+    // verify not claimed
+    let claimed = CLAIM.may_load(deps.storage, (&info.sender, U8Key::from(stage)))?;
+    if claimed.is_some() {
+        return Err(ContractError::Claimed {});
+    }
+
+    let config = CONFIG.load(deps.storage)?;
+    let merkle_root = MERKLE_ROOT.load(deps.storage, stage.into())?;
+
+    let user_input: String = info.sender.to_string() + &amount.to_string();
+    let mut hash: [u8; 32] = sha3::Keccak256::digest(user_input.as_bytes())
+        .as_slice()
+        .try_into()
+        .map_err(|_| ContractError::WrongLength {})?;
+
+    for p in proof {
+        let mut proof_buf: [u8; 32] = [0; 32];
+        hex::decode_to_slice(p, &mut proof_buf)?;
+
+        hash = match bytes_cmp(hash, proof_buf) {
+            Ordering::Less => sha3::Keccak256::digest(&[hash, proof_buf].concat())
+                .as_slice()
+                .try_into()
+                .map_err(|_| ContractError::WrongLength {}),
+            _ => sha3::Keccak256::digest(&[proof_buf, hash].concat())
+                .as_slice()
+                .try_into()
+                .map_err(|_| ContractError::WrongLength {}),
+        }?;
+    }
+
+    let mut root_buf: [u8; 32] = [0; 32];
+    hex::decode_to_slice(merkle_root, &mut root_buf)?;
+    if root_buf != hash {
+        return Err(ContractError::VerificationFailed {});
+    }
+
+    // Update claim index to the current stage
+    CLAIM.save(deps.storage, (&info.sender, stage.into()), &true)?;
+
+    let msgs: Vec<SubMsg> = vec![SubMsg::new(WasmMsg::Execute {
+        contract_addr: config.cw20_token_address.to_string(),
+        funds: vec![],
+        msg: to_binary(&Cw20ExecuteMsg::Transfer {
+            recipient: info.sender.to_string(),
+            amount,
+        })?,
+    })];
+    Ok(Response {
+        messages: msgs,
+        attributes: vec![
+            attr("action", "claim"),
+            attr("stage", stage),
+            attr("address", info.sender),
+            attr("amount", amount),
+        ],
+        data: None,
+        events: vec![],
+    })
+}
+
+fn bytes_cmp(a: [u8; 32], b: [u8; 32]) -> std::cmp::Ordering {
```

Comment:
This is basically how is ordering defined for arrayst= in Rust std: https://doc.rust-lang.org/std/primitive.array.html#impl-Ord, there is no point to redefine it.